src/extra/vui/vui.cpp

```diff
@@ -0,0 +1,243 @@
+#include "vui.hpp"
+#include "../gfx.hpp"
+#include "../inputs.hpp"
+#include "gfx/egui/egui_types.hpp"
+#include "gfx/linalg.hpp"
+#include "linalg_shim.hpp"
+#include "object_var_util.hpp"
+#include "params.hpp"
+#include "../egui/context.hpp"
+#include <gfx/renderer.hpp>
+#include <gfx/gizmos/gizmos.hpp>
+#include <input/input.hpp>
+
+namespace shards::VUI {
+
+#define VUI_PANEL_SHARD_NAME "VUI.Panel"
+
+struct VUIPanelShard;
+struct VUIContextShard {
+  VUIContext _vuiContext{};
+  SHVar *_vuiContextVar{};
+
+  Inputs::RequiredInputContext _inputContext;
+  gfx::RequiredGraphicsContext _graphicsContext;
+
+  ExposedInfo _exposedVariables;
+
+  std::vector<VUIPanelShard *> _panels;
+
+  input::InputBuffer _inputBuffer;
+
+  std::shared_ptr<gfx::GizmoRenderer> _debugRenderer;
+
+  PARAM_PARAMVAR(_queue, "Queue", "The draw queue to insert draw commands into.", {Type::VariableOf(gfx::Types::DrawQueue)});
+  PARAM_PARAMVAR(_view, "View", "The view that is being used to render.", {Type::VariableOf(gfx::Types::View)});
+  PARAM(ShardsVar, _contents, "Contents", "The list of UI panels to render.", {CoreInfo::ShardsOrNone});
+  PARAM_VAR(_scale, "Scale", "The scale of how many UI units per world unit.", {CoreInfo::FloatType});
+  PARAM_VAR(_debug, "Debug", "Render debug.", {CoreInfo::BoolType});
+  PARAM_IMPL(VUIContextShard, PARAM_IMPL_FOR(_queue), PARAM_IMPL_FOR(_view), PARAM_IMPL_FOR(_contents), PARAM_IMPL_FOR(_scale),
+             PARAM_IMPL_FOR(_debug));
+
+  VUIContextShard() {
+    _scale = Var(1000.0f);
+    _debug = Var(false);
+  }
+
+  static SHTypesInfo inputTypes() { return CoreInfo::NoneType; }
+  static SHTypesInfo outputTypes() { return CoreInfo::NoneType; }
```

Comment:
Try to add the corresponding `inputHelp()` and `outputHelp()`.

Also the idiomatic way for a shard that doesn't do anything with its input is to return it ("pass through"), in other words, `CoreInfo::AnyType`. Unless there is a reason here to make it a "black-hole"?

Reply:
Did passthrough